added possibility to define additional, non-gatsby schemas

kije opened the pull request:
This PR adds support for defining additional non-gatsby schemas as proposed in  #39

d4rekanguok commented:
Woah, thanks for taking this on!

kije commented:
@d4rekanguok PR is ready for review/merge

d4rekanguok commented:
Sorry about the lack of update here — I'm a bit occupied til the end of this week, but I'm really excited to check this PR out!

ricokahler commented:
> I'm a bit occupied til the end of this week

Same here actually 😢

If you want to speed things up, could you possibly add to our [`gatsby-starter-ts`](https://github.com/d4rekanguok/gatsby-typescript/tree/master/packages/gatsby-starter-ts) project in this PR?

We use that starter as final "end-to-end" test per se (it runs it our CI etc). If you're willing could you:

- add an additional schema pointing to some public API somewhere ([possibly GitHub's GraphQL API](https://developer.github.com/v4/) since it's stable)
- add a page that has a query (maybe written for Apollo) that uses this non-gatsby schema's codegen types

It doesn't have to be anything too exciting. We just need something that ensure this keeps working over time as dependencies update. Our unit tests can't really do that.

If you don't have time, I can definitely add these via another PR but I don't have bandwidth until this weekend unfortunately.

kije commented:
Hi, no problem. I added an additional schema to `gatsby-starter-ts` and implemented a simple query for it using `graphql-tag`. A minimal implementation and nothing fancy, but I think it does the job of providing an "end-to-end" for this feature.
The question is, of course, if you want the `gatsby-starter-ts` to include this sample implementation, since if someone else wants to use it as a starting point, they always need to remove this sample implementation ...

ricokahler commented on packages/gatsby-plugin-graphql-codegen/readme.md:
```diff
@@ -28,6 +28,16 @@ module.exports = {
 |options.fileName| `graphql-type.ts` | path to the generated file. By default, it's placed at the project root directory & it should not be placed into `src`, since this will create an infinite loop|
 |options.codegenDelay| `200` | amount of delay from file change to codegen|
 |options.pluckConfig| <pre>{ globalGqlIdentifierName: "graphql", modules: [ { name: 'gatsby', identifier: 'graphql' } ] }</pre> | options passed to [graphql-tag-pluck](https://github.com/ardatan/graphql-toolkit/tree/master/packages/graphql-tag-pluck) when extracting queries and fragments from documents |
+|options.additionalSchemas| <pre>[]</pre> | array of additional schemas (other that the schema used by gatsby queries) for which types should be generated for. This is useful e.g. when you use client-side queries (e.g. with apollo-client) where you are querying another schema/endpoint |
```
```suggestion
|options.additionalSchemas| <pre>[]</pre> | array of additional schemas (other than the schema used by gatsby queries) for which types should be generated for. This is useful when you use client-side queries (e.g. with apollo-client) where you are querying another schema/endpoint |
```

ricokahler commented on packages/gatsby-plugin-graphql-codegen/readme.md:
```diff
@@ -28,6 +28,16 @@ module.exports = {
+|fileName| graphql-types-${key}.ts | path to the generated file for this schema. By default, it's placed at the project root directory & it should not be placed into `src`, since this will create an infinite loop |
+|documentPaths| value of `options.documentPaths` | The paths to files containing graphql queries.  See also `options.documentPaths` |
+|pluckConfig| - | options passed to [graphql-tag-pluck](https://github.com/ardatan/graphql-toolkit/tree/master/packages/graphql-tag-pluck) when extracting queries and fragments from documents |
+|schema| - | additional schema to process. Can either be an url, a path to a local schema definition (both `.json` and `.graphql` are supported) or an inline definition. See also https://github.com/ardatan/graphql-toolkit#-schema-loading |
```
> Can either be an url, a path to a local schema definition (both `.json` and `.graphql` are supported) or an inline definition

💯

ricokahler commented on packages/gatsby-plugin-ts/readme.md:
```diff
@@ -61,6 +61,16 @@ In order for this plugin to work right, you'd need to set your compile options l
 |options.fileName| `graphql-type.ts` | path to the generated file. By default, it's placed at the project root directory & it should not be placed into `src`, since this will create an infinite loop|
 |options.codegenDelay| `200` | amount of delay from file change to codegen|
 |options.pluckConfig| <pre>{ globalGqlIdentifierName: "graphql", modules: [ { name: 'gatsby', identifier: 'graphql' } ] }</pre> | options passed to [graphql-tag-pluck](https://github.com/ardatan/graphql-toolkit/tree/master/packages/graphql-tag-pluck) when extracting queries and fragments from documents |
+|options.additionalSchemas| <pre>[]</pre> | array of additional schemas (other that the schema used by gatsby queries) for which types should be generated for. This is useful e.g. when you use client-side queries (e.g. with apollo-client) where you are querying another schema/endpoint |
```
I think this would be easier to maintain as just a link to `packages/gatsby-plugin-graphql-codegen/readme.md`. We can deep link with anchors (e.g. `thing/#anchor-id) for a better docs experience.

ricokahler commented on packages/gatsby-starter-ts/src/components/seo.tsx:
```diff
@@ -93,4 +94,37 @@ const SEO: React.FC<Props> = ({
   )
 }
 
+export const someTestQuery = gql`
```
I think it would be beneficial to add comment here explaining what this is.



```suggestion
/**
 * `gatsby-plugin-ts`/`gatsby-plugin-graphql-codegen` supports codegen for
 * non-gatsby schemas. This is useful if you're using a client-side GraphQL client.
 *
 * Read more at: https://github.com/d4rekanguok/gatsby-typescript/blob/master/packages/gatsby-plugin-graphql-codegen/readme.md#additional-schema-options-for-optionsadditionalschemas
 */
export const someTestQuery = gql`
```

ricokahler commented on packages/gatsby-plugin-graphql-codegen/src/gatsby-node.ts:
```diff
@@ -75,10 +158,10 @@ export const onPostBootstrap: NonNullable<GatsbyNode['onPostBootstrap']> = async
     }
     const { schema } = store.getState()
-    await buildDebounce(schema)
+    await buildDebounce(await prepareSchemas(schema, additionalSchemas))
```
does it make sense to prepare the schemas once before the build? I'm somewhat concerned about this line because of the await before the debounce. If there is a potential longer async task, I think it makes sense to debounce it.

so the way I see it, we have two options:

1. debounce `prepareSchemas` with `build`
2. prepare the schemas once before build (preferred & simpler)

What do you think?

cc @d4rekanguok

kije replied:
You're absolutely right, totally missed that, the `prepareSchema` call should be debounced as well in this case.

The idea here was to update the schemas before each codegen, so that if the schema updates, those changes are "instantaneously" reflected in the generated types (as it happens in the currently released version of the plugin if the gatsby schema changed during "runtime"). However, I think this is rarely the case and it complicates the implementation to do this right, so if you agree, I would change the implementation to only prepare the schemas once before build (option 2).

If at a later date it becomes clear that updating the schemas continuously is a requirement, it could be implemented as a new feature. Maybe then it could be implemented so that the schemas are updated in a fixed interval (like `gatsby-source-graphql` does with the `refetchInterval` option) instead of before every codegen.

ricokahler replied:
Yeah this is an interesting product question. I prefer option 2 too, I think it's simpler to stick to generating it once at startup but I can see it being annoying to have to restart the gatsby server if you change the schema but my bet is users don't dev both the client and server at the same time.

I'll let @d4rekanguok have the final word on that

d4rekanguok replied:
Yeah lets go with option 2!

kije replied:
Ok, great. I've just updated the implementation :+1:

ricokahler commented on packages/gatsby-plugin-graphql-codegen/src/gatsby-node.ts:
```diff
@@ -34,31 +51,97 @@ const getOptions: GetOptions = pluginOptions => ({
+          reporter.panic(
+            `[gatsby-plugin-graphql-codegen] schema ${key} does not exist!`
+          )
+        }
```
i don't think this will ever happen. should it be removed?

d4rekanguok commented on packages/gatsby-plugin-graphql-codegen/src/gatsby-node.ts:
```diff
@@ -75,10 +165,11 @@ export const onPostBootstrap: NonNullable<GatsbyNode['onPostBootstrap']> = async
   }
 
   // HACKY: might break when gatsby updates
   store.subscribe(watchStore)
```
IIRC these action types are not really accurate — it triggers on any file changes (not sure if that's still the case.)

kije replied:
I tried it out by running `gatsby develop` in  `gatsby-starter-ts`, changing the query for the additional schema a few times. Each time the changes I made triggered the codegen and the types were updated -> so this still works as it did before

d4rekanguok commented on packages/gatsby-plugin-graphql-codegen/src/gatsby-node.ts:
```diff
@@ -34,30 +51,103 @@ const getOptions: GetOptions = pluginOptions => ({
+  callback: (item: T, index: number, collection: T[]) => Promise<TResult>
+) => Promise<TResult[]>
+const asyncMap: AsyncMap = (collection, callback) =>
+  Promise.all(collection.map(callback))
```
I extract thing like this often 😅 IMO inline typescript definition is really hard to read inline, and I think breaking off code into smaller functions is also a way to improve readability (arguable), especially for utils functions. For those reasons, I would vote to keep it as is.

kije replied:
This function is basically a bit of syntax sugar.

I think this here is easier to read and understand
```typescript
const someArray = await asyncMap(values, async (el) => await someOperation(el))
```

than this

```typescript
const someArray = await Promise.all(values.map(async (el) => await someOperation(el)))
```

But if you want I can surely inline it :smile:

d4rekanguok commented:
Thanks for the review @ricokahler, always so thorough. It looks like we're really close, great work @kije!

kije commented:
@d4rekanguok @ricokahler Sorry for the long delay, I'm currently rather busy at work so I didn't have time to work on this PR. I hope I will be able to resolve the feedback this weekend.

d4rekanguok commented:
No worries @kije, whenever you feel like it!

kije commented on packages/gatsby-plugin-graphql-codegen/src/gatsby-node.ts:
```diff
@@ -34,30 +51,103 @@ const getOptions: GetOptions = pluginOptions => ({
+    },
+    ...additionalSchemas.map(({ schema, ...config }) => {
+      return {
+        fileName: `graphql-types-${config.key}.ts`,
```
@ricokahler `config` refers here to the [config object](https://github.com/kije/gatsby-typescript/tree/feature/additional_schemas/packages/gatsby-plugin-graphql-codegen#additional-schema-options-for-optionsadditionalschemas) of the additional schema, where the file name for the types of the additional schema might or might not be defined by the user of the plugin.

kije commented on packages/gatsby-plugin-ts/src/gatsby-node.ts:
```diff
@@ -4,6 +4,7 @@ import * as tsloader from 'ts-loader'
 import FTCWebpackPlugin from 'fork-ts-checker-webpack-plugin'
 import {
   onPostBootstrap as onPostBootstrapCodegen,
+  SchemaConfig,
```
No, I've removed this import

kije commented on packages/gatsby-plugin-graphql-codegen/src/gatsby-node.ts:
```diff
@@ -34,30 +51,103 @@ const getOptions: GetOptions = pluginOptions => ({
+const loadAdditionalSchema: LoadAdditionalSchema = schema =>
+  loadSchema(schema, {
+    loaders: [new UrlLoader(), new JsonFileLoader(), new GraphQLFileLoader()],
+  })
```
Done

kije commented on packages/gatsby-plugin-graphql-codegen/src/gatsby-node.ts:
```diff
@@ -34,30 +51,103 @@ const getOptions: GetOptions = pluginOptions => ({
+      fileName: initialConfig.fileName,
+      generateFromSchema: await generateWithConfig(initialConfig),
+    })),
+  ])
```
Done

kije commented on packages/gatsby-plugin-graphql-codegen/src/gatsby-node.ts:
```diff
@@ -34,30 +51,103 @@ const getOptions: GetOptions = pluginOptions => ({
+    })),
+  ])
+
+  const build = async (schemas: PreparedSchemas): Promise<void> => {
```
Done

kije commented:
@ricokahler I've resolved your feedback (note my replies to your comments above).
Please let me know if there is anything else I should change.

/cc @d4rekanguok

d4rekanguok commented:
This is awesome! Thanks again @kije @ricokahler 

Successfully published:
 - gatsby-plugin-graphql-codegen@2.6.0
 - gatsby-plugin-ts@2.6.0
